In [1]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
# 1.1.1 Split data into train and test data
from sklearn.model_selection import train_test_split

# 1.1.2 PReprocessong:
from sklearn.preprocessing import StandardScaler

# 1.1.3 Import class DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier as dtree

# 1.1.4 To draw decision tree
from sklearn.tree import export_graphviz
from sklearn.tree import plot_tree
import matplotlib.pyplot as plt
from sklearn.tree import export_graphviz
from sklearn.metrics import accuracy_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix
from sklearn import metrics
from sklearn.metrics import classification_report

from sklearn.datasets import load_iris
import os

from sklearn.model_selection import train_test_split

from sklearn.tree import plot_tree
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('Real fake News.csv')

In [3]:
df.head()

Unnamed: 0  Unnamed: 0.1               author  \
0           0             0         Jill Terreri   
1           1             1  Eleanor Hildebrandt   
2           2             2     Emily Zirkelbach   
3           3             3        Madeline Heim   
4           4             4        Sabine Martin   

                                           statement  \
0  New York is "leading the nation with the lowes...   
1  “Iowa ranks 45th in the U.S. for internet conn...   
2  “FUN FACT: West Virginia is home to zero billi...   
3  Says gubernatorial candidate Rebecca Kleefisch...   
4  The Biden administration "published a study co...   

                                              source  target BinaryTarget  \
0  NYS Department of Corrections and Community Su...    True         REAL   
1                                         Cindy Axne    True         REAL   
2                                         David Gura    True         REAL   
3                                   Kelda Helen Roys    True         REAL   
4                                     Randy Feenstra    True         REAL   

   BinaryNumTarget  
0                1  
1                1  
2                1  
3                1  
4                1

In [4]:
df.describe()

Unnamed: 0  Unnamed: 0.1  BinaryNumTarget
count  4870.000000   4870.000000      4870.000000
mean   1217.500000   1217.500000         0.500000
std     702.996251    702.996251         0.500051
min       0.000000      0.000000         0.000000
25%     609.000000    609.000000         0.000000
50%    1217.500000   1217.500000         0.500000
75%    1826.000000   1826.000000         1.000000
max    2435.000000   2435.000000         1.000000

In [5]:
data = df.values
print(data)

[[0 0 'Jill Terreri' ... True 'REAL' 1]
 [1 1 'Eleanor Hildebrandt' ... True 'REAL' 1]
 [2 2 'Emily Zirkelbach' ... True 'REAL' 1]
 ...
 [2433 2433 'Kiannah Sepeda-Miller' ... False 'FAKE' 0]
 [2434 2434 'Manuela Tobias' ... False 'FAKE' 0]
 [2435 2435 'Bill McCarthy' ... False 'FAKE' 0]]


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4870 entries, 0 to 4869
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Unnamed: 0       4870 non-null   int64 
 1   Unnamed: 0.1     4870 non-null   int64 
 2   author           4870 non-null   object
 3   statement        4870 non-null   object
 4   source           4870 non-null   object
 5   target           4870 non-null   bool  
 6   BinaryTarget     4870 non-null   object
 7   BinaryNumTarget  4870 non-null   int64 
dtypes: bool(1), int64(3), object(4)
memory usage: 271.2+ KB


In [7]:
df.isna().sum()

Unnamed: 0         0
Unnamed: 0.1       0
author             0
statement          0
source             0
target             0
BinaryTarget       0
BinaryNumTarget    0
dtype: int64

In [8]:
df.shape

(4870, 8)

In [9]:
df.head(10)

Unnamed: 0  Unnamed: 0.1               author  \
0           0             0         Jill Terreri   
1           1             1  Eleanor Hildebrandt   
2           2             2     Emily Zirkelbach   
3           3             3        Madeline Heim   
4           4             4        Sabine Martin   
5           5             5          Andy Nguyen   
6           6             6        Laura Schulte   
7           7             7          Emma Copley   
8           8             8         John Brumley   
9           9             9           D.L. Davis   

                                           statement  \
0  New York is "leading the nation with the lowes...   
1  “Iowa ranks 45th in the U.S. for internet conn...   
2  “FUN FACT: West Virginia is home to zero billi...   
3  Says gubernatorial candidate Rebecca Kleefisch...   
4  The Biden administration "published a study co...   
5  "Southwest Airlines announce they will no long...   
6  “The annual Pentagon budget hovers over $750 b...   
7  “If you look at the hospitalizations at Ruby M...   
8  “Overdose deaths in WV are up 45% from the pri...   
9  The GOP-controlled Legislature has "refused to...   

                                              source  target BinaryTarget  \
0  NYS Department of Corrections and Community Su...    True         REAL   
1                                         Cindy Axne    True         REAL   
2                                         David Gura    True         REAL   
3                                   Kelda Helen Roys    True         REAL   
4                                     Randy Feenstra    True         REAL   
5                                     Facebook posts    True         REAL   
6                                         Mark Pocan    True         REAL   
7                           Kathryn Moffett Bradford    True         REAL   
8                                     David McKinley    True         REAL   
9                                       Chris Larson    True         REAL   

   BinaryNumTarget  
0                1  
1                1  
2                1  
3                1  
4                1  
5                1  
6                1  
7                1  
8                1  
9                1

In [10]:
df.drop(['Unnamed: 0','Unnamed: 0.1','statement','target','BinaryNumTarget'], axis=1, inplace= True)

In [11]:
df.head()

author                                             source  \
0         Jill Terreri  NYS Department of Corrections and Community Su...   
1  Eleanor Hildebrandt                                         Cindy Axne   
2     Emily Zirkelbach                                         David Gura   
3        Madeline Heim                                   Kelda Helen Roys   
4        Sabine Martin                                     Randy Feenstra   

  BinaryTarget  
0         REAL  
1         REAL  
2         REAL  
3         REAL  
4         REAL

In [12]:
from sklearn.preprocessing import LabelEncoder

In [13]:
le = LabelEncoder()

In [14]:
df['author'] = le.fit_transform(df['author'])
df['source'] = le.fit_transform(df['source'])
df['BinaryTarget'] = le.fit_transform(df['BinaryTarget'])

In [15]:
from sklearn.preprocessing import StandardScaler

In [16]:
ss = StandardScaler()

In [17]:
x = df.iloc[:,0:2]
y = df.iloc[:,2]

In [18]:
ss.fit(x)

StandardScaler()

In [19]:
x=ss.transform(x)

In [20]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.3,random_state=1)

# KNN Classifier

In [21]:
knn = KNeighborsClassifier(n_neighbors=14, metric='euclidean')
knn.fit(x_train, y_train)

KNeighborsClassifier(metric='euclidean', n_neighbors=14)

In [22]:
y_pred = knn.predict(x_test)
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred)

0.78507871321013

# Decision Tree Classfier using Entropy

In [23]:
from sklearn.tree import DecisionTreeClassifier as dtree
ct = dtree(
           criterion="entropy",    # Alternative 'gini'
           max_depth=None       # Alternative, specify an integer
                              # 'None' means full tree till single leaf
           )
ct.fit(x_train,y_train)

DecisionTreeClassifier(criterion='entropy')

In [24]:
y_te = ct.predict(x_test)

In [25]:
accuracy_score(y_test,y_te)

0.8528405201916496

# Decision Tree using Gini

In [26]:
from sklearn.tree import DecisionTreeClassifier as dtree
ct = dtree(
           criterion="gini",   
           max_depth=None       # Alternative, specify an integer
                              # 'None' means full tree till single leaf
           )
ct.fit(x_train,y_train)

DecisionTreeClassifier()

In [27]:
y_tes = ct.predict(x_test)
accuracy_score(y_test,y_tes)

0.8528405201916496

# Random Forest Classifier

In [28]:
from sklearn.ensemble import RandomForestClassifier
clf=RandomForestClassifier(n_estimators=10)
clf.fit(x_train,y_train)

y_pred1=clf.predict(x_test)
accuracy_score(y_test,y_pred1)

0.8740588637919233

# KNN Accuracy : 0.78507871321013

# DTree Gini Accuracy : 0.8528405201916496

# DTree Entropy Accuracy: 0.8528405201916496

# Random Forest Accuracy: 0.8740588637919233

# Through the 4 accuracy tests we can see that the highest accuracy is using Random Forest, i.e., 87.40%